ENH JTFS `meta()`, `out_type='dict'`, `out_type='list'`

lostanlen opened the pull request:
Use of the `DryBackend` in JTFS ! (same as in #914)

Plus:
* `unpad_frequency`
* better `meta` information from `frequency_scattering` (including `n1_stride`)
* a new core JTFS function, `jtfs_average_and_format`, which takes care of the logic behind `average_global`, `time_averaging`, `frequency_averaging`, and `unpad_frequency`. Shortens and simplifies the `base_frontend`
* `out_type='dict'` and `out_type='list'` are now supported and tested. Coefficients in the dict / list are unpadded in frequency. (according to `n1_max` and `n1_stride` metadata)

Future work: 
* `format == 'time'` while require a new primitive, `backend.split`. Right now it raises a `NotImplementedError` in `jtfs_average_and_format`. This is next

lostanlen commented:
That's a lot of misses. Will add more unit-tests later today

cyrusvahidi commented on kymatio/scattering1d/backend/numpy_backend.py:
```diff
@@ -121,4 +121,9 @@ def pad_frequency(cls, x, padding):
 
+    @classmethod
+    def unpad_frequency(cls, x, n1_max, n1_stride):
+        n1_unpadded = 1 + (n1_max//n1_stride)
```
`n1_unpadded = 1 + (n1_max // n1_stride)`

svp

lostanlen replied:
29b1990388d548b96518303d9ec7bc94743c3ec6

changhongw commented on kymatio/scattering1d/frontend/base_frontend.py:
```diff
@@ -175,12 +175,11 @@ def meta(self):
-        class DryBackend:
-            __getattr__ = lambda self, attr: (lambda *args: None)
-
+        backend = self._DryBackend()
```
Why introduce `backend` here? It is used only once in line 181. Why not directly use `self._DryBackend()` in line 181, similar to lines 596 and 600?

lostanlen replied:
a41de1b6dafcbcb9d00bbc8d2cb1d82cd54b8df7

lostanlen commented:
ok, i have addressed feedback and have written a tet for `meta()`. 

there will be a few misses, corresponding to the `format='time'`case. PR incoming

ready for another round of code review

cyrusvahidi commented:
TODO 
* port "unpad_frequency" to PyTorch and TF backends
* write tensorflow frontend
* test tensorflow differentiability
* write keras frontend
* write sklearn frontend
* write backend.split in all backends
* support out_type='list' in base frontend

cyrusvahidi commented on kymatio/scattering1d/frontend/base_frontend.py:
```diff
@@ -594,99 +593,155 @@ def scattering(self, x):
+        meta['spin'] = np.sign(meta['xi_fr'])
+        return meta
+
+    def meta(self):
```
why duplicate meta?

cyrusvahidi replied:
I've added a commit that removes the duplicate definition of `meta`

lostanlen replied:
oops! this is definitely a rebase issue. thanks

lostanlen commented:
@cyrusvahidi ready for merge?

lostanlen commented:
@cyrusvahidi yes i approve. but formally speaking, you and changhong are the reviewers in this. 
<img width="343" alt="Screenshot 2022-08-30 at 15 35 35" src="https://user-images.githubusercontent.com/3943142/187451245-ea57167c-3948-4eb1-8731-98f363d531df.png">

next time you want a review from me, please summon me via the review panel

lylyhan commented on tests/scattering1d/test_timefrequency_scattering.py:
```diff
@@ -6,35 +6,39 @@
     assert (jtfs._N_padded_fr - N_input_fr) > (2 * jtfs.F)
-    assert (jtfs._N_padded_fr - N_input_fr)  > (2 * (2 ** jtfs.J_fr))
-    phi_f = gauss_1d(2 * jtfs._N_padded_fr, jtfs.sigma0/jtfs.F)
+    assert (jtfs._N_padded_fr - N_input_fr) > (2 * (2**jtfs.J_fr))
```
is this necessary? if there is a previous line checking jtfs.F = 2**jtfs.J_fr

lostanlen replied:
ah you're right. this is redundant

lylyhan commented on tests/scattering1d/test_timefrequency_scattering.py:
```diff
@@ -69,7 +73,7 @@ def test_check_runtime_args():
     assert S._check_runtime_args() is None
```
the S is not initialized anywhere before this line
